proposals/list-patterns.md

```diff
@@ -0,0 +1,122 @@
+# List patterns
+
+## Summary
+
+Lets you to match an array or a list with a sequence of patterns e.g. `array is {1, 2, 3}` will match an integer array of the length three with 1, 2, 3 as its elements, respectively.
+
+## Detailed Design
+
+The pattern syntax is modified as follow:
+
+```antlr
+recursive_pattern
+  : type? positional_pattern_clause? length_pattern_clause? property_or_list_pattern_clause? simple_designation?
+  ;
+
+length_pattern_clause
+  : '[' pattern ']'
+  ;
+
+property_or_list_pattern_clause
+  : list_pattern_clause
+  | property_pattern_clause
+  ;
+
+property_pattern_clause
+  : '{' (subpattern (',' subpattern)* ','?)? '}'
+  ;
+
+list_pattern_clause
+  : '{' pattern (',' pattern)* ','? '}'
+  ;
+
+slice_pattern
+  : '..' negated_pattern?
+  ;
+
+primary_pattern
+  : recursive_pattern
+  | slice_pattern
+  ;
+```
+There are two new additions to the *recursive_pattern* syntax as well as a *slice_pattern*:
+
+- The *list_pattern_clause* is used to match elements and the *length_pattern_clause* is used to match the length.
+- A *slice_pattern* is only permitted once and only directly in a *list_pattern_clause* and discards _**zero or more**_ elements.
+
+Due to the ambiguity with *property_pattern_clause*, the *list_pattern_clause* cannot be empty and a *length_pattern_clause* should be used instead.
+
+#### Pattern compatibility
+
+A *length_pattern_clause* is compatible with any type that conforms to the following rules:
+
+1. Has an accessible property getter that returns an `int` and has the name `Length` or `Count`
+
+A *list_pattern_clause* is compatible with any type that conforms to the following rules:
+
+1. Is compatible with the length pattern
+2. Has an accessible indexer with a single `int` parameter
+
+A *slice_pattern* is compatible with any type that conforms to the following rules:
+
+1. Is compatible with the list pattern
```

Comment:
Do you mean "length" pattern here?
Let's try to re-use the terminology from the ranges spec: "The type is Countable".

What if the type has an `this[Range]` indexer?

Reply:
`this[Range]` is not part of range spec. It just "accidentally" supported via overload resolution and `Range` natural type for `..` expressions.

If we want to support it, we should explicitly add a section.

Reply:
Let's confirm, but my assumption is that we'd want to support `this[Range]`. 
The implicit Range indexing was added to make the transition easier (legacy types didn't have Range indexer). That was not called out in the range spec, because it already worked (normal indexing).